### PR TITLE
[Merged by Bors] - Rate limit RPC requests

### PR DESCRIPTION
**Pull request by divagant-martian**
## Issue Addressed
#1056 

## Proposed Changes
- Add a rate limiter to the RPC behaviour. This also means the rate limiting occurs just before the door to the application level, so the number of connections a peer opens does not affect this (this would happen in the future if put on the handler)
- The algorithm used is the leaky bucket as a meter / token bucket implemented the GCRA way
- Each protocol has its own limit. Due to the way the algorithm works, the "small" protocols have a hard limit, while bbrange and bbroot allow [burstiness](https://www.wikiwand.com/en/Burstiness). This is so that a peer can't request hundreds of individual requests expecting only one block in a short period of time, it also allows a peer to send two half size requests instead of one with max if they want to without getting limited, and.. it also allows a peer to request a batch of the maximum size and then send _appropriately spaced_ requests of really small sizes. From what I've seen in sync this is plausible when reaching the target slot.

## Additional Info
Needs to be heavily tested

### CODE REVIEW

File: beacon_node/eth2_libp2p/src/rpc/mod.rs [hunk 1]
```diff
@@ -97,7 +102,25 @@ pub struct RPC<TSpec: EthSpec> {
 impl<TSpec: EthSpec> RPC<TSpec> {
     pub fn new(log: slog::Logger) -> Self {
         let log = log.new(o!("service" => "libp2p_rpc"));
+        let limiter = RateLimiterBuilder::new()
+            .one_every(Protocol::MetaData, Duration::from_secs(5))
+            .one_every(Protocol::Ping, Duration::from_secs(15))
```
Comment (AgeManning) [inline beacon_node/eth2_libp2p/src/rpc/mod.rs:107]:
```suggestion
            .one_every(Protocol::Ping, Duration::from_secs(5))
```

I think Teku ping us every 5 seconds, so lets allow them

File: beacon_node/eth2_libp2p/src/rpc/mod.rs [hunk 2]
```diff
@@ -97,7 +102,25 @@ pub struct RPC<TSpec: EthSpec> {
...
+            .one_every(Protocol::Status, Duration::from_secs(1))
```
Comment (AgeManning) [inline beacon_node/eth2_libp2p/src/rpc/mod.rs:108]:
```suggestion
            .n_every(Protocol::Status, 3, Duration::from_secs(15))
```
Hopefully we dont get status'd too often. We sometimes do double status's so maybe we'll allow some burst for status

File: beacon_node/eth2_libp2p/src/rpc/mod.rs [hunk 3]
```diff
@@ -193,13 +216,47 @@ where
...
+                    // send the event to the user
+                    self.events
+                        .push(NetworkBehaviourAction::GenerateEvent(RPCMessage {
+                            peer_id,
+                            conn_id,
+                            event,
+                        }))
+                }
+                Err(RateLimitedErr::TooLarge) => {
+                    // we set the batch sizes, so this is a coding/config err
```
Comment (AgeManning) [inline beacon_node/eth2_libp2p/src/rpc/mod.rs:232]: The protocol should reject these before they reach here right? We should never hit this.

File: beacon_node/eth2_libp2p/src/rpc/mod.rs [hunk 4]
```diff
@@ -193,13 +216,47 @@ where
...
+                }
+                Err(RateLimitedErr::TooSoon(until)) => {
+                    warn!(self.log, "Request exceeds the rate limit";
+                        "request" => req.to_string(), "peer_id" => peer_id.to_string(), "earliest_in_ms" => until.as_millis());
+                    let event = Err(HandlerErr::Inbound {
+                        id: *id,
+                        proto: req.protocol(),
+                        error: RPCError::RateLimited,
+                    });
+                    self.events
```
Comment (AgeManning) [inline beacon_node/eth2_libp2p/src/rpc/mod.rs:244]: Can we also respond to the peer with an Error, telling them they got rate limited, also in the case `TooLarge`

Other clients were discussing sending back the `until` also, but I don't think we support that yet in the spec.

File: beacon_node/eth2_libp2p/src/rpc/rate_limiter.rs [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        let tau = self.tau;
+        let t = self.t;
+        // how long does it take to replenish these tokens, when at least one will be taken
+        let additional_time = t * (tokens - 1);
+        if additional_time + t > tau {
+            // the time required to process this amount of tokens is longer than the time that
+            // makes the bucket full. So, this batch can _never_ be processed
+            return Err(RateLimitedErr::TooLarge);
+        }
+        // If the peer is new, we consider their bucket full
```
Comment (AgeManning) [inline beacon_node/eth2_libp2p/src/rpc/rate_limiter.rs:225]: When new peers connect they immediately request probably Status, Ping and Metadata. 
Are we going to instantly reject these with a rate limit?

File: beacon_node/eth2_libp2p/src/rpc/mod.rs (hunk 4 above)
Comment (divagant-martian) [reply beacon_node/eth2_libp2p/src/rpc/mod.rs]: I sent (as a string) how much they should wait. If we then need the actual until, it won't be hard

**Pull request closed without merging**